spine/model/layer/cnn/ppn.py

```diff
@@ -484,9 +492,75 @@ def process_loss_config(self, mask_loss='CE', resolution=5.,
         else:
             raise ValueError(
                     f"Mask loss name not recognized: {mask_loss}")
+            
+            
+    @staticmethod
+    def get_ppn_positives(coords : torch.Tensor, 
+                          labels: torch.Tensor, 
+                          ppn_labels: torch.Tensor, 
+                          resolution: float,
+                          offset: int):
+        """Get ppn positive label mask, but with cluster restrictions.
+
+        Parameters
+        ----------
+        coords : torch.Tensor
+            N x 3 3D coordinates of the image voxels
+        labels : torch.Tensor
+            N x 1 tensor of the particle id label for each voxel
+        ppn_labels : torch.Tensor
+            N_p x 6 PPN label points 
+        resolution : float
+            Distance from a label point in pixels within which a voxel is
+            
+        Warning
+        -------
+        If a given cluster has two ppn label points (tracks), then they must
+        appear consecutively in the ppn_labels tensor.
+
+        Returns
+        -------
+        positives : torch.Tensor
+            N x 1 tensor of the positive label mask
+        """
+        positives = torch.zeros(coords.shape[0], device=coords.device, dtype=torch.bool)
+        closests = -torch.ones(coords.shape[0], device=coords.device, dtype=torch.long)
+        
+        # test_mask = torch.zeros(coords.shape[0], device=coords.device, dtype=torch.bool)
+        
+        # print("Labels = ", labels.unique(return_counts=True), ppn_labels[:, PPN_LPART_COL])
+        
+        frag_ids = set()
```

Comment:
Why do we need to play this game? We know that there is either one or two points per particle, depending whether it is a shower or a track, so I believe all points should be relevant? Why do we need to keep track of particle IDs we have already seen?

Reply:
This is because the ppn code needs to keep track of the index of the closest ppn truth point for every point in `coords`. Here we loop over all `ppn_labels`. If the point `pt` is a shower ppn label, then in fact every point inside that shower fragment will point to the same "closest ppn truth point", since we restrict point-label correspondence to within fragments. However if the point `pt` is a track, then it will have two ppn label points (start and end). In that case, when asked about the "closest ppn truth point", roughly half of the points in that track cluster will point to the startpoint and the other half to the endpoint. This is what's done here:
```python
        # Loop over all ppn label points
        for i, pt in enumerate(ppn_labels):
            frag_id = int(pt[PPN_LPART_COL])
            if frag_id in frag_ids:
            # This means this is the second ppn label point for tracks. 
            # Since this track was already accounted for, we skip 
                continue
            frag_ids.add(frag_id)
            mask = labels == frag_id
            # Get 3D coordinates for start (and end point, if exists). 
            label_points = ppn_labels[ppn_labels[:, 5] == frag_id][:, COORD_COLS] # Either (1,3) or (2,3) tensor
            dist_mat = local_cdist(coords[mask], label_points)
            # Within coords[mask], keep track of the closest label ppn point's index
            min_return = torch.min(dist_mat, dim=1) # Will be a 0-filled tensor or a [0,1]-filled tensor
            # Closest indices are given by offset (index offset due to batching) 
            # + i (this ppn label point's index within this batch, in ppn_labels)
            # + min_return.indices (for showers, all 0's; for tracks, 0 or 1 indicating point is closer to start vs. end)
            closests[mask] = offset + (i + min_return.indices)
            positives_frag = (dist_mat < resolution).any(dim=1)
            positives[mask] = positives_frag
```

Reply:
Ok that makes sense, but you can just check the `shape` of the PPN label point for that right? I don't think you need to keep track of the `frag_ids` in a `set`. Also, not a fan of `fragment`/`frag`; these are all particles (in the LArCV definition of the term). Fragment implies there's been another breaking step (e.g. DBSCAN).

Reply:
I think if you use `shape` to check if the ppn label points are showers or tracks (1 vs 2), then in the next loop you would have to check if the two points you are currently seeing have been already accounted for?

Reply:
Forget about `shape`, why are you even looping over label points instead of particles? Can't you just loop over unique particle indexes? Something like:

```
part_ids = ppn_labels[:, PPN_LPART_COL]
for part_id in np.unique(part_ids):
    point_index = np.where(part_ids == part_id)[0]
    index = np.where(labels == part_id)[0]
    points = ppn_labels[point_index][:, COORD_COLS]
    dist_mat = local_cdist(coords[index], points)
    positives[index] = (dist_mat < resolution).any(dim=1)
    min_return = torch.min(dist_mat, dim=1)
    closest[index] = points_index[min_return.indices]
```
with comments, of course.

Also just noticed your code above has a hardcoded `5` for column access, which is a big no no.

And where is `offset` even set or updated? I removed it.

Reply:
`offset` is not set or updated but is an argument of the function and is necessary for keeping track of global index vs. within-batch index (for `closest`)

Reply:
Ok that makes sense, does the rest of the comment make sense though?

Reply:
But otherwise this change would work and is more concise--I will change the code accordingly

Reply:
Thank you!